tempto-core/src/main/java/com/teradata/tempto/s3/S3Client.java

```diff
@@ -0,0 +1,384 @@
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.ObjectListing;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.slf4j.LoggerFactory.getLogger;
+import static java.nio.charset.Charset.defaultCharset;
+
+/**
+ * S3 client based on the Amazon AWS client API
+ */
+public class S3Client
+{
+    private static final Logger logger = getLogger(S3Client.class);
+
+    public static final String CONF_S3_AUTH_KEY_ID = "s3.auth.keyid";
+    public static final String CONF_S3_AUTH_SECRET = "s3.auth.secret";
+    public static final String CONF_S3_BUCKET_NAME = "s3.bucketname";
+
+    private final String bucketName;
+    private final AmazonS3 s3client;
+
+    /**
+     * Constructor
+     *
+     * @param keyId         AWS access key ID
+     * @param secret        AWS secret access key
+     * @param bucketName    S3 bucket to work in
+     */
+    public S3Client(
+            @Named(CONF_S3_AUTH_KEY_ID) String keyId,
+            @Named(CONF_S3_AUTH_SECRET) String secret,
+            @Named(CONF_S3_BUCKET_NAME) String bucketName)
+    {
+        // Check null values and set object variables
+        checkNotNull(keyId, "aws api access key id is null");
+        checkNotNull(secret, "aws api secret access key is null");
+        this.bucketName = checkNotNull(bucketName, "s3 bucket name is null");
+
+        // Initialize AWS SDK credentials
+        BasicAWSCredentials credentials = null;
+        try {
+            credentials = new BasicAWSCredentials(keyId, secret);
+        } catch (Exception e) {
+            throw new AmazonClientException("Cannot initialize credentials", e);
+        }
+
+        // Create S3 client
+        client = AmazonS3ClientBuilder.standard()
+                .withCredentials(new AWSStaticCredentialsProvider(credentials))
+                .build();
+    }
+
+    /**
+     * Method to get folder contents
+     *
+     * @param path  path to get contents
+     * @return      list of object names
+     */
+    public ArrayList<String> getFolderContents(String path)
+    {
+        return getObjects(path);
+    }
+
+    /**
+     * Method to get folder contents (including folders)
+     *
+     * @param path              path to get contents
+     * @param includeFolders    true/false to include folders or not
+     * @return
+     */
+    public ArrayList<String> getFolderContents(String path, boolean includeFolders)
+    {
+        ArrayList<String> objects = new ArrayList<String>();
+
+        // If no forward slash at end of path, add one
+        if (!path.substring(path.length() - 1).equals("/")) {
+            path += "/";
+        }
+
+        for (String object : getAllObjects(path)) {
+            // Remove input path from object name and determine the number of forward slashes left
+            String objectName = object.replace(path, "");
```

Comment:
Adn what if `path` appears in `object` two times.
E.g. input path path is `/foo` and
we have `/foo/foo/foo` on S3.

Generally this logic is very hard to follow.
I suggest that you rewrite logic here and operate on path segments.
Use guava's `Splitter.on('/')` to split input path and results of `getAllObjects` to cut paths int parts.
Then for each entity coming from `getAllObjects` take N-th element (where N is number of segments in input path) and collapse everything to `Set`.